### ISSUE TITLE
Sign in user & get user object - Facebook & Google Login - AWS Amplify + React Native

### ISSUE THREAD

**Issue by kevwang19** [aws-amplify/amplify-js#4621]
Currently building a react native mobile app to use federated/social logins FB and Google with AWS Amplify. Followed some tutorials (mostly this one: https://dev.to/dabit3/the-complete-react-native-guide-to-user-authentication-with-the-amplify-framework-ib2) and amplify docs in setting up the login process, and am able to get to a certain point - will use FB login as an example.

1. Used Auth.federatedSignIn({ provider: "Facebook" }) provided by amplify docs to a button on my login screen
2. Can see on my Cognito user pool console that a user was created through FB
3. Callback url on Cognito app client settings successfully redirects screen back to my mobile app
4. At this point, used Auth.currentAuthenticatedUser() provided by amplify to try to get user, but keep on getting not authenticated / user not signed in 

So basically my problem is that I can't get the user to be signed in after Cognito redirects to my react native app despite the user being created to the cognito user pool. I tried the hosted UI (would rather use my custom UI) solution as well, but getting the same issue of retrieving user after being redirected to my app, despite user having been created in user pool. Any help or insight appreciated!

**Comment by Amplifiyer** [comment]
@kevwang19 can you provide your full sample app code here? Also have you tried using `withOAuth` [HOC](https://aws-amplify.github.io/docs/js/authentication#react-components)

**Comment by kernwig** [comment]
Looks like the same problem I'm facing with an Angular app. Upon sign-in with Google, the app reloads and I see the Cognito user information in local storage. Yet, `Auth.authStateChange$` reports a state of `“signedOut”`. Explicitly requesting the current auth user also returns `“No current user”.`  If I reload the page or navigate to the page with the amplify-authenticator component, then the state becomes `“signedIn”` and all is well.

Source code:
Call to `federatedSignIn`: https://github.com/kernwig/sqac-amplify/blob/468b8816aadb4c0ba5c1b80101d92909a8627b33/src/app/home/account.component.ts

Watch of `Auth.authStateChange$`: https://github.com/kernwig/sqac-amplify/blob/468b8816aadb4c0ba5c1b80101d92909a8627b33/src/app/services/user.service.ts

Live app: https://d3l0j9nusq7n6r.cloudfront.net/#/home/account

**Comment by kevwang19** [comment]
@Amplifiyer, I've seen online that more people have had success with the HOC, but it limits customizability, and I prefer to just use the social sign-ins. Below is my code for the login page:

`const SwitchNav = createSwitchNavigator({
  Login: {
    screen: Login
  },
  AppNavigator: {
    screen: AppNavigator
  }
})

const AppContainer = createAppContainer(SwitchNav);

export default class App extends React.Component {
  componentDidMount() {
    this.checkAuth().then(()=> console.log('loaded compnonent, checked auth'))
  }
  checkAuth = async () => {
    try {
      await Auth.currentAuthenticatedUser()
      .then((user)=>{console.log(user)})
      .then(()=>{
        this.navigator.dispatch(
          NavigationActions.navigate({ routeName: 'AppNavigator' })
        )
      })

    } catch (err) {
      this.navigator.dispatch(
        NavigationActions.navigate({ routeName: 'Login' })
      )
      console.log('there is no user')
    }
  }

  render() {
    return (
      <AppContainer
        ref={nav => this.navigator = nav}
        onNavigationStateChange={this.checkAuth}
      />
    )
  }
}
`

On the login screen, I just have a button that has onPress={() => Auth.federatedSignIn({ provider: "Facebook" })}

If I press that button, it takes me to a safari browser to sign in through facebook, which then redirects through aws, then back to my app. It's just that when I get back to my app to check the authentication state, it still returns not authenticated. I'm thinking maybe it's because of the order in which the functions are called...perhaps the checkAuth functions are executing before the federated signin auth flow has a chance to return the signed in user.

In any case, I actually ended up using the react-native fbsdk instead, and got that to work instead of going through amazon cognito.
This example at the bottom works with FBSDK: 
https://github.com/magus/react-native-facebook-login/issues/176

**Comment by kevwang19** [comment]
I guess also I didn't quite understand the onNavigationStateChange part in the dev.to/dabit3 tutorial I mentioned earlier... seems to me that in order to call the function checkAuth, there needs to be a navigation state change, but wouldn't that be the result, not the trigger of the function?

**Comment by mynameisguy** [comment]
I'm having the same issue in react-native app.
Users get created but after I log in, calling `Auth.currentAuthenticatedUser();` return an error.
Regular sign in with username and password works (running `Auth.currentAuthenticatedUser();`  returns the user).

**Comment by Amplifiyer** [comment]
> I'm thinking maybe it's because of the order in which the functions are called...perhaps the checkAuth functions are executing before the federated signin auth flow has a chance to return the signed in user.

@kevwang19 I believe that could be the root cause. Check the full React example [here](https://aws-amplify.github.io/docs/js/authentication#react-components) for capturing hub events. Ideally you will sign in and wait for the hub events to fire a `signIn` event which you can listen and then call `checkUser`. As an example, this is what HOC does https://github.com/aws-amplify/amplify-js/blob/master/packages/aws-amplify-react-native/src/Auth/withOAuth.js#L70

**Comment by kernwig** [comment]
@Amplifiyer, given that the same problem happens with Angular, and the solution you point to is pretty much what I'm unsuccessfully attempting.

Either the same bug exists for both React and Angular, or in code shared by both.

It appears that this process only works if the login component is on the landing page after the redirect back to the application.

**Comment by Amplifiyer** [comment]
@kernwig, yes if you are using the HOCs then the landing page after the redirect should have the component since it's parsing the URL and updating the auth state. In your component, you can listen to the Hub events and after a successful signIn event, redirect the User to whatever landing page you desire.

**Comment by kernwig** [comment]
Okay, found two fixes based on your comments @Amplifiyer, both involving using the `Hub` to listen for `signIn`. I first use this as a direct analogue to monitoring `Auth.authStateChange$`. The `Hub` told me when a federated user signed in/out, and `authStateChange$` for a user pool user. I still wanted to land on my accounts page though. So I changed it to navigate to the account page on `Hub` `signIn` event. Doing that caused `Auth.authStateChange$` to trigger when the auth component rendered, so I just rely on that. Here's the code change: https://github.com/kernwig/sqac-amplify/commit/8de758e5dcc8c0f6ef7ce729aa872acf0d96abc2

Note: The `Hub` doesn't play nice with Angular, thus the need to use `NgZone` directly.

**Comment by Amplifiyer** [comment]
@kernwig thanks for the update. @kevwang19 let us know if you are still facing any issues.
